### PR TITLE
Fix metadata handling and object loading for new map format

### PR DESCRIPTION
**Pull request by ihhub**
- fix a crash while removing monsters from a map without changing their properties
- fix a crash while loading a map with random castles
- add code to handle all needed metadata for objects
- add extra checks in debug build for map loading
- fix Abandoned Mines update after capturing
- fix inability to select any players except the one placed on a map

Relates to #6845

### CODE REVIEW

**Review by Districh-ru — CHANGES_REQUESTED** [review]
Hi, @ihhub, I left several comments. Could you please check them when you have time?

There is also an issue: the Random Castles are not correctly updated because there is no `BUILD_CASTLE` built. I'll fix it in the Castle Details PR.

File: src/fheroes2/world/world_loadmap.cpp [hunk 1]
```diff
@@ -692,56 +694,93 @@ bool World::loadResurrectionMap( const std::string & filename )
...
 
                 Castle * castle = new Castle( static_cast<int32_t>( tileId ) % width, static_cast<int32_t>( tileId ) / width, race );
-                const uint8_t color = Maps::getTownColorIndex( map, tileId, object.id );
-                castle->SetColor( 1 << color );
+                castle->SetColor( color );
 
                 vec_castles.AddCastle( castle );
 
+                if ( isRandom ) {
+                    Maps::UpdateCastleSprite( castle->GetCenter(), castle->GetRace(), castle->isCastle(), true );
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/world/world_loadmap.cpp:736]: @ihhub, we can do a fix for Random Castles update by replaceing `castle->isCastle()` with `townObjects[object.index].objectType == MP2::OBJ_RANDOM_CASTLE`:
```suggestion
                    Maps::UpdateCastleSprite( castle->GetCenter(), race, castle->townObjects[object.index].objectType == MP2::OBJ_RANDOM_CASTLE, true );
```

[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/world/world_loadmap.cpp:747]: Hi @Districh-ru , I've fixed it another way :)

File: src/fheroes2/world/world_loadmap.cpp [hunk 2]
```diff
@@ -692,56 +694,93 @@ bool World::loadResurrectionMap( const std::string & filename )
...
+                const int color = ( 1 << Maps::getTownColorIndex( map, tileId, object.id ) );
+
+                int race = ( 1 << townObjects[object.index].metadata[0] );
+                const bool isRandom = ( race == Race::RAND );
+
+                if ( isRandom ) {
+                    assert( townObjects[object.index].objectType == MP2::OBJ_RANDOM_CASTLE || townObjects[object.index].objectType == MP2::OBJ_RANDOM_TOWN );
+
+                    const Kingdom & kingdom = GetKingdom( color );
+                    race = static_cast<uint8_t>( kingdom.GetRace() );
```
Comment (Districh-ru) [inline src/fheroes2/world/world_loadmap.cpp]: For the neutral player we should pick a really random race. :)
```suggestion
                    race = ( color == Color::NONE ) ? Race::Rand() : static_cast<uint8_t>( kingdom.GetRace() );
```

**Review by Districh-ru — APPROVED** [review]
👍

**Pull request merged**